core/src/main/java/org/apache/iceberg/avro/PruneColumns.java

```diff
@@ -90,12 +92,18 @@ public Schema record(Schema record, List<String> names, List<Schema> fields) {
       } else if (fieldSchema != null) {
         hasChange = true;
         filteredFields.add(copyField(field, fieldSchema, fieldId));
+      } else if (emptyStructIds.contains(fieldId)) {
+        // This field does not require any known sub-fields but is required in the projection so keep it without
```

Comment:
This is the pruning Side fix to Bug 1, Basically making sure we add the parents of any fields in the projection even if none of their real (in the file) fields are projected.

For example without this fix

```
From (A, B { C } )
Project (A, B { D.Optional })
Returns
(A) 
```